gravitee-am-service/src/main/java/io/gravitee/am/service/impl/PasswordHistoryService.java

```diff
@@ -0,0 +1,138 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.am.service.impl;
+
+import io.gravitee.am.common.audit.EventType;
+import io.gravitee.am.identityprovider.api.User;
+import io.gravitee.am.model.PasswordHistory;
+import io.gravitee.am.model.PasswordSettings;
+import io.gravitee.am.model.ReferenceType;
+import io.gravitee.am.repository.management.api.PasswordHistoryRepository;
+import io.gravitee.am.service.AuditService;
+import io.gravitee.am.service.authentication.crypto.password.PasswordEncoder;
+import io.gravitee.am.service.exception.PasswordHistoryException;
+import io.gravitee.am.service.reporter.builder.AuditBuilder;
+import io.gravitee.am.service.reporter.builder.management.UserAuditBuilder;
+import io.reactivex.Maybe;
+import io.reactivex.Single;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Component;
+
+import java.util.Date;
+import java.util.List;
+
+import static java.util.Comparator.comparing;
+
+/**
+ * Service providing password history.
+ */
+@Component
+public class PasswordHistoryService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(PasswordHistoryService.class);
+    private final PasswordHistoryRepository repository;
+    private final AuditService auditService;
+    private final PasswordEncoder passwordEncoder;
+
+    @Autowired
+    public PasswordHistoryService(@Lazy PasswordHistoryRepository repository, AuditService auditService, PasswordEncoder passwordEncoder) {
+        this.repository = repository;
+        this.auditService = auditService;
+        this.passwordEncoder = passwordEncoder;
+    }
+
+    /**
+     * Checks that a password has not already been used in the allotted history and adds it if it has not.
+     * Returns a {@link PasswordHistory} instance on completion.
+     *
+     * @param referenceType    the reference type (DOMAIN, APPLICATION)
+     * @param referenceId      id of the reference
+     * @param user           id of user for this password
+     * @param rawPassword      unencrypted password provided by the user. Passed separate from the user object as its password is nulled after creation to avoid leakage.
+     * @param principal        user performing this action
+     * @param passwordSettings domain/application password settings
+     * @return Single containing {@link PasswordHistory} or an error if the password was already in the history.
+     */
+    public Maybe<PasswordHistory> addPasswordToHistory(ReferenceType referenceType, String referenceId, io.gravitee.am.model.User user, String rawPassword, User principal, PasswordSettings passwordSettings) {
+        LOGGER.debug("Adding password history entry for user {}", user);
+        if (passwordSettings == null || !passwordSettings.isPasswordHistoryEnabled()) {
```

Comment:
In SCIM password may be missing, we should also return a Maybe.empty if rawPassword is null. 
Or do not call this method into the upstream service if the password is null.

Reply:
Doesn't the code return an error when the password is empty?

`// check password
                                if (isInvalidUserPassword(userToUpdate.getPassword(), userToUpdate)) {
                                    return Single.error(new InvalidValueException("Field [password] is invalid"));
                                }`

Reply:
Not for an update and I have open an issue about that. 
Password is an optional field on SCIM spec and as we can declare a user linked to a social IDP, the password isn't required in this case.